File: BOJ/1405.cpp
#include <stdio.h>
#include <algorithm>
using namespace std;
int gox[4] = {0,0,1,-1 };
int goy[4] = {1,-1,0,0 };
bool visited[33][33];
double answer;
int n;
double point[4];
// 단순경로라함은 간곳은 방문하지않는다. 즉 사이클이 아니다.
void dfs(int x, int y, int depth, double po) {
	visited[x][y] = true;
	if (depth == n) {
		answer += po;
		return;
	}
	for (int i = 0; i < 4; i++) {
		int nx = x + gox[i];
		int ny = y + goy[i];
		if (!visited[nx][ny]) {
			dfs(nx, ny, depth + 1, po*point[i]);
			visited[nx][ny] = false; //갔다온곳을 해제해줘야하므로
		}
	}
}
int main() {
	scanf("%d", &n);
	for (int i = 0; i < 4; i++) {
		scanf("%lf", &point[i]);
		point[i] /= 100;
	}
	dfs(15, 15, 0, 1.0);
	printf("%.9lf\n", answer);
	return 0;
}

File: BOJ(Java 공부용)/1405.java
import java.io.*;
import java.util.StringTokenizer;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int[] gox = {0, 0, 1, -1};
    static int[] goy = {1, -1, 0, 0};
    static double ans = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        double[] arr = new double[4];
        boolean[][] visited = new boolean[33][33];
        for (int i = 0; i < 4; i++) {
            arr[i] = Double.parseDouble(st.nextToken());
            arr[i] /= 100;
        }
        dfs(15, 15, 0, n, 1.0, visited, arr);
        System.out.print(String.format("%.9f", ans));
    }

    private static void dfs(int x, int y, int cnt, int depth, double sum, boolean[][] visited, double[] arr) {
        visited[x][y] = true;
        if (cnt == depth) {
            ans += sum;
            return;
        }
        for (int i = 0; i < 4; i++) {
            int nx = x + gox[i], ny = y + goy[i];
            if (!visited[nx][ny]) {
                dfs(nx, ny, cnt + 1, depth, sum * arr[i], visited, arr);
                visited[nx][ny] = false;
            }
        }
    }

}